### PR TITLE
[#156804143] Upgrade CF from v1.21.0 to v1.28.0

### PR DESCRIPTION
**Pull request by bandesz**
## What

Upgrade cf-deployment from v1.21.0 to v1.28.0

We're upgrading CF as part of the regular platform update. See story at [1].

For a cf-deployment diff see [2].

Please note that the buildpacks will be upgraded in a separate story as we
usually notify our tenants a week ahead.

We also upgraded the stemcells to 3468.30 and 3541.12.

### Deployment manifest changes

* garden-runc-release: `garden.graph_cleanup_threshold_in_mb=0` was removed

    From the spec file:

    ```
    garden.graph_cleanup_threshold_in_mb:
        description: "DEPRECATED in favour of grootfs.reserved_space_for_other_jobs_in_mb, unless also using deprecated_use_garden_shed property."
        default: -1
    ```

    We don't set `deprecated_use_garden_shed` so this change is fine.

* garden-runc-release: `grootfs.reserved_space_for_other_jobs_in_mb=15360` - this is the default value in the spec file.

### Release changes

```
$ git diff v1.21.0..v1.28.0 cf-deployment.yml | grep '[+-] *url' | sed 's|bosh.io/d/||;s|?v=|/releases/tag/v|' | grep -v "buildpack"
-  url: https://github.com/cloudfoundry/capi-release/releases/tag/v1.52.0
+  url: https://github.com/cloudfoundry/capi-release/releases/tag/v1.54.0
-  url: https://github.com/cloudfoundry/cf-mysql-release/releases/tag/v36.11.0
+  url: https://github.com/cloudfoundry/cf-mysql-release/releases/tag/v36.12.0
-  url: https://github.com/cloudfoundry/cf-networking-release/releases/tag/v1.12.0
+  url: https://github.com/cloudfoundry/cf-networking-release/releases/tag/v1.13.0
-  url: https://github.com/cloudfoundry/cf-smoke-tests-release/releases/tag/v40.0.1
+  url: https://github.com/cloudfoundry/cf-smoke-tests-release/releases/tag/v40.0.2
-  url: https://github.com/cloudfoundry/cf-syslog-drain-release/releases/tag/v6.0
+  url: https://github.com/cloudfoundry/cf-syslog-drain-release/releases/tag/v6.4
-  url: https://github.com/cloudfoundry/cflinuxfs2-release/releases/tag/v1.194.0
+  url: https://github.com/cloudfoundry/cflinuxfs2-release/releases/tag/v1.203.0
-  url: https://github.com/cloudfoundry/diego-release/releases/tag/v2.1.0
+  url: https://github.com/cloudfoundry/diego-release/releases/tag/v2.5.0
-  url: https://github.com/cloudfoundry/garden-runc-release/releases/tag/v1.12.1
+  url: https://github.com/cloudfoundry/garden-runc-release/releases/tag/v1.13.1
-  url: https://github.com/cloudfoundry-incubator/cf-routing-release/releases/tag/v0.174.0
+  url: https://github.com/cloudfoundry-incubator/cf-routing-release/releases/tag/v0.175.0
-  url: https://github.com/cloudfoundry/statsd-injector-release/releases/tag/v1.2.0
+  url: https://github.com/cloudfoundry/statsd-injector-release/releases/tag/v1.3.0
-  url: https://github.com/cloudfoundry/uaa-release/releases/tag/v56
+  url: https://github.com/cloudfoundry/uaa-release/releases/tag/v57.1
```

Additionally the routing-release version was 0.172.0, so we need to ugprade to 0.175.0.

### Significant changes

* routing-release: Operators can now configure the manifest property router.sanitize_forwarded_proto: true to sanitize the X-Forwarded-Proto HTTP header in a request when router.force_forwarded_proto_https is set to false (default value). We recommend setting the property to true if the Gorouter is the first component to terminate TLS, and setting it to false when your load balancer is terminating TLS and setting the X-Forwarded-Proto header details (default: false) - we can leave this as false as we terminate TLS before the gorouter in haproxy

* cf-syslog-drain-release: Add scalablesyslog.adapter.max_bindings property to limit the number of bindings that can be serviced by an adapter. Default 500. Important Note This feature will reject new streams and could result in syslog drain impact for users if your deployment is not scaled properly. See [3] to our adapter scaling recommendations for guidance. - we might want to add some monitors to see early if we need to scale our adapters and this value.

[1] https://www.pivotaltracker.com/n/projects/1275640/stories/156804143
[2] https://github.com/cloudfoundry/cf-deployment/compare/v1.21.0...v1.28.0
[2] https://github.com/cloudfoundry/cf-syslog-drain-release#syslog-adapter

## How to review

1. Validate that we created the following branches from the right versions:
    * https://github.com/alphagov/paas-cf-smoke-tests/tree/gds-cf-deployment-v1.28.0
    * https://github.com/alphagov/paas-cf-acceptance-tests/tree/cf1.28-allow-http-redirect (this contains https://github.com/cloudfoundry/cf-acceptance-tests/pull/290)

1. Enable datadog in paas-bootstrap and run the create-bosh-concourse pipeline

```
cd paas-bootstrap
SELF_UPDATE_PIPELINE=false ENABLE_DATADOG=true make dev deployer-concourse pipelines
```

1. Enable datadog in paas-cf and run create-cloudfoundry from this branch. All tests should pass.

```
SELF_UPDATE_PIPELINE=false ENABLE_DATADOG=true BRANCH=upgrade-cf-156804143 make dev pipelines
```

1. We ran the credential rotation and it succeeded without errors - We don't think it's worth repeating

1. ❗️ The certificate rotation is currently broken (see https://www.pivotaltracker.com/story/show/157032509), so we don't test that.

## Who can review

Not @henrytk or me.

### CODE REVIEW

**Comment by keymon** [comment]
In https://www.cloudfoundry.org/blog/usn-3619-2/ it is recommended to upgrade to 3541.12. 

Shall we do it in this PR?

**Comment by bandesz** [comment]
@keymon already did it as part of this PR

**Comment by keymon** [comment]
> @keymon already did it as part of this PR

:+1:

File: manifests/cf-manifest/spec/manifest/release_versions_spec.rb [hunk 1]
```diff
@@ -20,4 +20,27 @@
         "expected release #{release['name']} version #{release['version']} to have matching version in URL: #{release['url']}"
     end
   end
+
+  specify "manifest versions are newer than the ones in cf_deployment" do
```
Comment (alext) [inline manifests/cf-manifest/spec/manifest/release_versions_spec.rb]: Nitpick: this is actually asserting that they're `>=`, whereas the wording here suggests it's asserting that they're `>`.

File: manifests/cf-manifest/spec/manifest/release_versions_spec.rb [hunk 2]
```diff
@@ -20,4 +20,27 @@
...
+    }.to_h
+
+    cf_deployment_manifest.fetch("releases").each do |release|
+      next if release['name'].end_with? '-buildpack'
+
+      if manifest_releases.has_key? release['name']
+        cf_deployment_release_version = release.fetch('version')
+        manifest_release_version = manifest_releases[release['name']]
+
+        puts "#{normalise_version(manifest_release_version)} >= #{normalise_version(cf_deployment_release_version)}"
```
Comment (alext) [inline manifests/cf-manifest/spec/manifest/release_versions_spec.rb]: Was this meant to be left in? It looks like a debugging thing.

File: manifests/cf-manifest/spec/manifest/release_versions_spec.rb [hunk 3]
```diff
@@ -20,4 +20,27 @@
...
+        expect(normalise_version(manifest_release_version)).to be >= normalise_version(cf_deployment_release_version),
+          "expected release #{release['name']} version #{manifest_release_version} to be older than #{cf_deployment_release_version} as defined in cf-deployment"
```
Comment (alext) [inline manifests/cf-manifest/spec/manifest/release_versions_spec.rb]: This failure message is backwards. eg it currently outputs something like:
> expected release routing version 0.172.0 to be older than 0.175.0 as defined in cf-deployment

I think it's missing a "not" in there.

**Comment by alext** [comment]
Looks good. It applies cleanly to my dev environment.

There have been some new commits to both the smoke tests and acceptance tests repos since this was cut, but I don't think any of those changes apply to us, so I don't think it's worth updating.

I think this is ready to go once the comments above are addressed.

**Comment by bandesz** [comment]
The manifest versions test was fixed.

**Pull request merged**